File: tree/binary-tree-maximum-path-sum.cpp
class Solution {
public:
    int path = INT_MIN;

    int findpath(TreeNode* node) {
        if (node == nullptr)
            return 0;

        // Ignore negative paths
        int left = max(0, findpath(node->left));
        int right = max(0, findpath(node->right));

        // Update global max path
        path = max(path, left + right + node->val);

        // Return max single path sum
        return node->val + max(left, right);
    }

    int maxPathSum(TreeNode* root) {
        findpath(root);
        return path;
    }
};

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        traversal(root);
        return max;
    }

    int traversal(TreeNode node){
        if (node == null){
            return 0;
        }

        int left = Math.max(0, traversal(node.left));
        int right = Math.max(0, traversal(node.right));
        max = Math.max(max, node.val + left + right);

        return node.val + Math.max(left, right);
    }
}